lib/spack/docs/packaging_guide.rst

```diff
@@ -4054,36 +4054,193 @@ File functions
+   from the set of executables)
+
+If extra attributes are returned, they will be recorded in ``packages.yaml``
+and be available for later reuse. As an example, the ``gcc`` package will record
+by default the different compilers found and an entry in ``packages.yaml``
+would look like:
+
+.. code-block:: yaml
+
+   packages:
+     gcc:
+       externals:
+       - spec: 'gcc@9.0.1 languages=c,c++,fortran'
+         prefix: /usr
+         compilers:
+           c: /usr/bin/x86_64-linux-gnu-gcc-9
+           c++: /usr/bin/x86_64-linux-gnu-g++-9
+           fortran: /usr/bin/x86_64-linux-gnu-gfortran-9
+
+This permits, for instance, to keep track of executables that would be named
+differently if built by Spack (e.g. ``x86_64-linux-gnu-gcc-9``
+instead of just ``gcc``).
+
+"""""""""""""""""""""""""""
+Filter matching executables
+"""""""""""""""""""""""""""
+
+Sometimes defining the appropriate regex for the ``executables``
+attribute might prove to be difficult, especially if one has to
+deal with corner cases or exclude "red herrings". To help keeping
+the regular expressions as simple as possible, each package can
+optionally implement a ``filter_executables`` method:
+
+.. code-block:: python
+
+    @classmethod
+    def filter_detected_exes(cls, prefix, exes_in_prefix):
+        """Return a filtered list of the executables in prefix"""
+
+which takes as input a prefix and a list of matching executables and
+returns a filtered list of said executables.
+
+Using this method has the advantage of allowing custom logic for
+filtering, and does not restrict the user to regular expressions
+only.  Consider the case of detecting the GNU C++ compiler. If we
+try to search for executables that match ``g++``, that would have
+the unwanted side effect of selecting also ``clang++`` - which is
+a C++ compiler provided by another package - if present on the system.
+Trying to select executables that contain ``g++`` but not ``clang``
+would be quite complicated to do using regex only. Employing the
+``filter_detected_exes`` method it becomes:
+
+.. code-block:: python
+
+   class Gcc(Package):
+      executables = ['g++']
+
+      def filter_detected_exes(cls, prefix, exes_in_prefix):
+         return [x for x in exes_in_prefix if 'clang' not in x]
+
+Another possibility that this method opens is to apply certain
+filtering logic when specific conditions are met (e.g. take some
+decisions on an OS and not on another).
+
+^^^^^^^^^^^^^^^^^^
+Validate detection
+^^^^^^^^^^^^^^^^^^
+
+To increase detection robustness, packagers may also implement a method
+to validate the detected Spec objects:
+
+.. code-block:: python
+
+   @classmethod
+   def validate_detected_spec(cls, spec, extra_attributes):
+       """Validate a detected spec. Raise an exception if validation fails."""
+
+This method receives a detected spec along with its extra attributes and must
+raise an exception of any type if the spec is considered invalid. In that case
```

Comment:
We certainly could. What would be the gain though? The code that deals with that is:

https://github.com/spack/spack/blob/b52c236a2ca7843042e6049c6e06596aacb2b9aa/lib/spack/spack/cmd/external.py#L279-L288

and I think we would still need to catch and log any `Exception` to make external detection resilient to errors in packages. As I see it right now, we might impose another constraint on users and what we gain is to log the reason why a spec was discarded with a different message. Not sure if this is worth it. Thoughts?

Reply:
> and I think we would still need to catch and log any Exception to make external detection resilient to errors in packages. 

That makes sense to me. I originally thought (based on the function parameters) that users wouldn't be calling filesystem functions etc. and so the only errors raised in this function would be explicitly raised by the user. However, it's true that we don't want an error in a single package to prevent package detection in general.

My request in this case is that users can also return `False` to indicate that the provided spec is not valid (so they don't have to think about what exception to raise).

Reply:
> My request in this case is that users can also return False to indicate that the provided spec is not valid (so they don't have to think about what exception to raise).

I was implementing this solution, but went back on it as it introduces some asymmetry: users can return `False` or raise, but they are never required to return `True` (and if they are we have a more complex API). I think the concern that users don't have to decide what to raise is a good one so I went with another solution in 150e477.

The idea is that the documentation now says explicitly that users should assert conditions on the spec (thus they don't even have to know that this operation raises an exception). The implementation stays the same as it is now permitting us to catch whatever is going wrong and proceed with the next spec / package. 

Since I went with a different solution I'll leave this discussion open. Let me know if this seems a better approach than before or if you want to revisit it.

Reply:
I think we may want to be prepared for user confusion over this: `assertions` are a slightly different paradigm than if/else/raise. I want to avoid users having to swap between paradigms while implementing packages.

That being said, assertions are OK, but I think it's also worth providing an exception type for users to raise, and to document that in case they want to do things like:

```
if spec_doesnt_meet_my_condition:
  raise NewSpackExceptionIntroducedInThisPR
```

Reply:
Done in 17b48bf Let me know if that suffices.